checks: Add various detections specific to Linux

mihawk90 opened the pull request:
### Description
This adds various detections aimed at helping us in linux-support to provide help quicker without having to dive deep into the log with every user. This should cover the most of the common questions and issues, and the myriad of possible combinations encountered on Discord.

Changes in detail:

* renamed `wayland.py` to `linux.py` since this isn't just for Wayland anymore
* moved the snap detection merged in #137 to `linux.py`
* for X11 vs. Wayland
  * useful for advice on the various capture methods
  * `CRITICAL` XWayland was already logged
  * `INFO` Wayland was previously already detected, but not logged
  * `INFO` X11 Logging + recommended captures
  * `WARNING` Also logs PipeWire capture sources being used on X11. Currently this is only relevant for very recent GNOME versions and for recent(-ish) KDE versions. KDE doesn't actually support this either, but still advertises availability.
  * `CRITICAL` On Wayland this also logs when no PipeWire captures are available
* `INFO` Log usage of the Flatpak package
  * helps to know this is an official package and is also relevant for plugin installation
* `INFO` Distro + Version if available (unfortunately not on Flatpak)
  * useful to advise on package managers and names for xdg-desktop-portals or plugin packages
* `INFO` Desktop Environment
  * useful for xdg-desktop-portals and the various limitations of specific DEs
* `INFO` Missing default plugins
  * currently checks for Browser, VLC, and WebSocket but can easily be extended if need be
  * while AJA and BlackMagic plugins are technically a default functionality, these are essentially never asked for so I didn't see much reason to include them
* `INFO` Missing v4l2loopback module for VCam functionality

### Motivation and Context
The current lack of output by the OBS Bot requires us to manually check for the same information on every log we are getting in linux-support. The detections and help texts (suggestions welcome!) should give users some pointers to help themselves when there is currently no supporter around (since linux-support isn't as frequented as the others). Obviously it also helps supporters to give faster responses without having to check every log manually.

Of course Linux has an inherent multitude of possible setups, which I tried to cover as best as possible without checking for every eventuality. It should cover the most frequent issues and questions collected over quite some time.

### How Has This Been Tested?
Grabbed a random assortment of logs from the Discord:
```sh
#!/bin/bash

echo "===== Snap ====="
./loganalyzer.py --url "https://obsproject.com/logs/a8VNcRHD4vjHol3P"

echo "===== Flatpak ====="
./loganalyzer.py --url "https://obsproject.com/logs/MXdrIL75xKFyPNom"

echo "===== Wayland ====="
./loganalyzer.py --url "https://obsproject.com/logs/MXdrIL75xKFyPNom"

echo "===== XWayland ====="
./loganalyzer.py --url "https://obsproject.com/logs/1FaZYJAAwOsKvUxX"

echo "===== Wayland + no pipewire ====="
./loganalyzer.py --url "https://obsproject.com/logs/CN9BBx9uPPjBcGlA"

echo "===== X11 ====="
./loganalyzer.py --url "https://obsproject.com/logs/jBUCYqPE5LRHMRHB"

echo "===== X11 + PipeWire sources ====="
./loganalyzer.py --url "https://obsproject.com/logs/MQhnzM1TFjBgdxrO"
```

### Types of changes
- New feature (non-breaking change which adds functionality)
- Tweak (non-breaking change to improve existing functionality)
- Documentation (a change to documentation pages)

### Checklist:
- [ ] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

RytoEX reviewed (changes requested):
If a commit changes one file in `checks/`, the commit prefix is that file's name without the file extension.

Always capitalize the first word in the commit message subject after the prefix.

I don't use Linux, so I'll probably need someone else to comment on the substance of the changes themselves.

mihawk90 commented:
Fixed up.

Also added a note regarding the missing default plugins in OP.

mihawk90 commented:
Adjustments for unified PipeWire source in 30.2:
- Added new `pipewire-screen-capture-source` detection
- renamed "Screen Capture" to "Display Capture" throughout the help texts

mihawk90 commented:
Rebased to current master to keep it up to date.

mihawk90 commented:
Rebased to current master and also fixed the name for the VLC plugin being wrong.

RytoEX commented:
I still need someone familiar with Linux or Linux support to give some insight here.

RytoEX commented on checks/linux.py:
```diff
@@ -0,0 +1,176 @@
+
+    if len(isFlatpak) > 0:
+        return [LEVEL_INFO, "Flatpak",
+                "You are using the Flatpak. Plugins are available as Flatpak extensions, which you can find in your Distributions Software Center or via <code>flatpak search com.obsproject.Studio</code>. Installation of external plugins is not supported."]
```
```suggestion
                "You are using the Flatpak. Plugins are available as Flatpak extensions, which you can find in your Distribution's Software Center or via <code>flatpak search com.obsproject.Studio</code>. Installation of external plugins is not supported."]
```

Or `distribution's`.

mihawk90 replied:
Fixed

RytoEX commented on checks/linux.py:
```diff
@@ -0,0 +1,176 @@
+        return [LEVEL_WARNING, "PipeWire capture on X11",
+                """While technically possible, most Desktop Environments do not implement the PipeWire capture portals on X11. Your captures will therefore likely not work, i.e. you will be unable to pick a window or display, or the selected source will stay empty.<br><br>
+                We generally recommend using \"Window Capture (Xcomposite)\" on X11, as \"Display Capture (XSHM)\" can introduce bottlenecks depending on your setup.<br><br>
+                From a technical standpoint a Desktop Environment should not advertise the availability of these captures if they are not implemented. This behaviour can therefore be considered a bug in your Desktop Environment."""]
```
```suggestion
                From a technical standpoint a Desktop Environment should not advertise the availability of these captures if they are not implemented. This behavior can therefore be considered a bug in your Desktop Environment."""]
```

mihawk90 replied:
Fixed

RytoEX commented on checks/linux.py:
```diff
@@ -0,0 +1,176 @@
+            Global Keyboard Shortcuts are not currently available under Wayland."""]
+
+
+def checkX11(lines):
```
Perhaps `checkX11Captures` instead? I'd have suggested `checkPipewireOnX11`, but this function also seems to make recommendations around non-PipeWire captures.

mihawk90 replied:
Renamed. Initially I had just named it `checkX11` because its already existing counterpart was `checkWayland` so I just kept it for consistency, the other checks came later :sweat_smile:

RytoEX commented on checks/linux.py:
```diff
@@ -0,0 +1,176 @@
+                We generally recommend using \"Window Capture (Xcomposite)\" on X11, as \"Display Capture (XSHM)\" can introduce bottlenecks depending on your setup.<br><br>
+                From a technical standpoint a Desktop Environment should not advertise the availability of these captures if they are not implemented. This behaviour can therefore be considered a bug in your Desktop Environment."""]
+
+    return [LEVEL_INFO, "X11",
+            "Window Capture is available via Xcomposite, while Display Capture is available via XSHM. We generally recommend sticking to \"Window Capture (Xcomposite)\" since \"Display Capture (XSHM)\" can introduce bottlenecks depending on your setup."]
```
Should this be emitted all the time even if the user is not using any captures?

mihawk90 replied:
Since this is only `INFO` this was intended, yes. We often get questions of what capture method to use, and there's not always someone around.

RytoEX replied:
I have mixed feelings about always emitting an INFO like this, but I don't have enough experience in Linux support to accurately weigh the benefits vs. the noise.

mihawk90 replied:
If it helps, it's not much different then Windows logs always showing the Windows version and Admin status. It's a helper info for Supporters (although not being active in w-s I don't know how much it helps or what the differences are), and for endusers reading the analysis it doesn't provide much detail either.

The big difference being that on Linux there's more variables to consider (Distro, Desktop Environment, and Window System mostly). So yes we end up with 3 info bulletins instead of 2, but it is information we look at in every log anyway. So it takes a click off supporters and if noone's around it might help the occasional user. If not, well then they'll have to wait for someone to come around anyway 😅

RytoEX replied:
Always showing the Windows version is "fine", which is why I'm okay with showing the Distro info elsewhere for Linux (though it might be nice if they could be a single item "system info"). This is mostly because we're aware of issues that only exist or are fixed in specific OS versions that are otherwise hard to diagnose from logged info.

The point I'm making here is that we're emitting an INFO giving blanket capture advice whether or not any such sources are detected. For all we know, the user won't be using any of the captures mentioned (maybe they're only using v4l2 or teleport or browser).

mihawk90 replied:
Yeah I get what you're saying.
I just honestly don't really see it being problematic.
If a user doesn't want to capture either windows or a display, they can just ignore it since it's clearly about window and display capture. Maybe we could reword it a little:

> If you wish to capture a window or an entire display, captures are available via Xcomposite and XSHM respectively. We generally recommend sticking to "Window Capture (Xcomposite)" since "Display Capture (XSHM)" can introduce bottlenecks depending on your setup.

That should make it even clearer that it can and should be ignored if no captures are intended.

If users do end up getting confused by it we can remove it anytime as well. We won't know unless we have it first though.

And going back to the windows logs, the same can be said about the "Not Admin" factoid, which can lead users to assume they should be running it as admin (similar to the audio buffering factoid in the other PR).

RytoEX commented on checks/linux.py:
```diff
@@ -0,0 +1,176 @@
+        return
+
+    modulesMissingList = []
+    modulesCheckList = ('obs-browser.so', 'obs-websocket.so', 'vlc-video.so')
```
What's the rationale for checking only these modules?

mihawk90 replied:
I was trying to find a balance between checking "everything" and those that might reasonably be missing and are actually asked for in support. This PR was also opened before the CMake rewrite hit master so a lot of things that might be disabled by default now used to be enabled by default so I figured it wouldn't make much sense to include.

Specifically for this list:
Most distros don't ship Browser by default, many don't ship Websocket due to missing dependencies, and VLC source requires VLC to be installed (and isn't available in Flatpak at all). So those are the 3 major ones that could be missing.
As noted in the OP I left out AJA and BlackMagic because they are virtually never asked for.

It just occurred to me that `obs-x264.so` might be useful to check for, although it's really only relevant for Fedora at the moment (the plugin is shipped in RPM Fusion not in Fedora repos).

That being said, I could check for all modules if desired.

RytoEX replied:
I don't have a strong opinion on this. I just wanted to raise a question.

RytoEX commented on checks/linux.py:
```diff
@@ -0,0 +1,176 @@
+    if len(hasV4L2Module) > 0:
+        return [LEVEL_INFO, "VCam not available",
+                """Using the Virtual Camera requires the <code>v4l2loopback</code> kernel module to be installed.<br>
+                If required, please refer to our <a href="https://github.com/obsproject/obs-studio/wiki/install-instructions#prerequisites-for-all-versions">Install Instructions</a> on how to install this on your distribution.<br>
```
Perhaps we should link to https://obsproject.com/kb/virtual-camera-troubleshooting instead (which also links to https://github.com/obsproject/obs-studio/wiki/install-instructions#prerequisites-for-all-versions).

RytoEX replied:
I don't have a strong opinion. This is better than no message.

RytoEX commented on checks/linux.py:
```diff
@@ -0,0 +1,176 @@
+        return [LEVEL_INFO, "VCam not available",
+                """Using the Virtual Camera requires the <code>v4l2loopback</code> kernel module to be installed.<br>
+                If required, please refer to our <a href="https://github.com/obsproject/obs-studio/wiki/install-instructions#prerequisites-for-all-versions">Install Instructions</a> on how to install this on your distribution.<br>
+                If the module was not already loaded OBS will normally ask you for permission to load it when required. This requires a working <code>polkit</code> setup.<br>
```
```suggestion
                If the module was not already loaded, OBS will normally ask you for permission to load it when required. This requires a working <code>polkit</code> setup.<br>
```

mihawk90 replied:
Fixed

Penwy commented on checks/linux.py:
```diff
@@ -0,0 +1,176 @@
+
+    distro = isDistroNix[0].split()
+    # Snap Package logs "Ubuntu Core" as distro, so it gets split halfway
+    if distro[2] == '"Ubuntu' and distro[3] == 'Core"':
```
I realise this is not the PR that introduced this, but relying on split() and precise array indices is not highly reliable, and more importantly, will crash the whole analyzer if split() returns less than 4 elements (which would happen in case of for example a log straight from sdtout so without timestamps).
I'd instead advocate for `if  'Ubuntu Core' in isDistroNix[0]:`

mihawk90 replied:
> and more importantly, will crash the whole analyzer if split() returns less than 4 elements (which would happen in case of for example a log straight from sdtout so without timestamps).

I tested this, and no:
```
 cat stdout.log | grep "Distribution"
info: Distribution: "Ubuntu Core"
^0    ^1            ^2      ^3
 ./loganalyzer.py -f stdout.log | grep "Ubuntu"
Info:     Beta OBS Version (31.0.0-beta2), No Output Session, Third-Party Plugins (1), "Ubuntu Core", X11, KDE (KDE), VCam not available
"Ubuntu Core"
```

On `stdout` there's an `info:` prefix in place of the timestamp, so the index count stays the same.

mihawk90 replied:
Fixed

Penwy commented on checks/linux.py:
```diff
@@ -0,0 +1,176 @@
+    if not sessionTypeLine:
+        return
+
+    sessionType = sessionTypeLine.split()[3]
+    if sessionType != 'wayland':
+        return
+
+    if len(isDistroNix) > 0:
+        distro = isDistroNix[0].split()
+        if distro[2] == '"Ubuntu"' and distro[3] == '"20.04"':
+            return [LEVEL_CRITICAL, "Ubuntu 20.04 under Wayland",
+                    "Ubuntu 20.04 does not provide the needed dependencies for OBS to capture under Wayland.<br> So OBS is able to capture only under X11/Xorg."]
```
Same as above, shouldn't rely on specific array indices

mihawk90 replied:
Fixed

Penwy reviewed (commented):
Besides the use of split and indices, I agree with all additions and can speak to their usefulness for linux support.

tiberium-v commented:
I think also they are useful. Also I have the feeling that we have more user in #linux-support, they don't like to follow the instructions for uploading the log and paste it via pastebin or co in. For me, who is sometimes on mobile, it is a pain to access it. So a more verbose log analyzer output is here helpful.

In general, I have the feeling, that the windows side has always more useful information in the analyzer output. Like the windows version/build. So the additions here with info about the linux distro/flatpak and which Display-Server is running is on my side welcome as someone who give support there.

RytoEX commented:
> overall its fine, im not sure I agree with trying to explain which captures to use on your platform for every log uploaded but we can always change it later if it becomes distracting to users.

That was the point I was trying to make in [this review thread](https://github.com/obsproject/loganalyzer/pull/161#discussion_r1854742820).

mihawk90 commented:
Since there were no objections to the [altered wording in the review thread](https://github.com/obsproject/loganalyzer/pull/161#discussion_r1855328918) above I went ahead and pushed it.

Let me know if you want further changes :smile:

RytoEX commented:
Can someone provide sample logs that would cause the INFO Wayland and INFO X11 items to be emitted?

I am tempted to ask that we remove the INFO Wayland and INFO X11 items from this PR so we can land everything else, and then we can consider those two INFO items in a separate PR.

RytoEX commented on checks/linux.py:
```diff
@@ -0,0 +1,176 @@
+    hasV4L2Module = search('v4l2loopback not installed', lines)
+
+    if len(hasV4L2Module) > 0:
+        return [LEVEL_INFO, "VCam not available",
```
```suggestion
        return [LEVEL_INFO, "Virtual Camera not available",
```

Spell out "Virtual Camera" unless we use "VCam" somewhere else in a user-facing message.

mihawk90 replied:
Changed. I kept it short because I wasn't sure whether the OBS Bot response has any length limitations for the bullet points.

mihawk90 commented:
> Can someone provide sample logs that would cause the INFO Wayland and INFO X11 items to be emitted?

Log examples are in the PR description, 

> I am tempted to ask that we remove the INFO Wayland and INFO X11 items from this PR so we can land everything else, and then we can consider those two INFO items in a separate PR.

I'm not opposed to removing one or both of the helptexts, however I would strongly advocate for keeping the bullet point itself. This is one of the things we need to check in almost every log, so having it in the discord bot's response would be a major timesaver. For most questions/issues users come into the support chat for the answer depends on whether the user is running on X11 or Wayland. It was one of the things that kicked this PR off in the first place.

mihawk90 commented:
Removed the helptexts for now. If we decide we want them back we can just revert 03a2708.

As noted above, I'm leaving the bulletpoints in as the display server is crucial information in linux-support.

Fenrirthviti commented on checks/linux.py:
```diff
@@ -0,0 +1,172 @@
+
+    if '"Ubuntu Core"' in isDistroNix[0]:
+        return [LEVEL_WARNING, "Snap Package",
+                "You are using the Snap Package. This is a community-supported modified build of OBS Studio; please file issues on the <a href=\"https://github.com/snapcrafters/obs-studio/issues\">Snapcrafters GitHub</a>.<br><br>OBS may be unable to assist with issues arising out of the usage of this package and therefore recommends following our <a href=\"https://obsproject.com/download#linux\"> ...
```
```suggestion
                "You are using the Snap Package. This is a community-supported modified build of OBS Studio; please file issues on the <a href=\"https://github.com/snapcrafters/obs-studio/issues\">Snapcrafters GitHub</a>.<br><br>OBS may be unable to assist with issues arising out of the usage of this package. We recommend following our <a href=\"https://obsproject.com/download#linux\">Install Instructions</a> instead."]
```

mihawk90 replied:
Done.

Fenrirthviti commented on checks/linux.py:
```diff
@@ -0,0 +1,172 @@
+    if len(isDistroNix) > 0:
+        if '"Ubuntu" "20.04"' in isDistroNix[0]:
+            return [LEVEL_CRITICAL, "Ubuntu 20.04 under Wayland",
+                    "Ubuntu 20.04 does not provide the needed dependencies for OBS to capture under Wayland.<br> So OBS is able to capture only under X11/Xorg."]
```
```suggestion
                    "Ubuntu 20.04 does not provide the needed dependencies for OBS to capture under Wayland.<br> Capture  will only function under X11/Xorg."]
```

mihawk90 replied:
Done.
God bless 20.04 is going EOL in 4 months :pray:

Fenrirthviti commented on checks/linux.py:
```diff
@@ -0,0 +1,172 @@
+
+    if (len(hasPipewireCaptureDesktop) > 0) or (len(hasPipewireCaptureWindow) > 0) or (len(hasPipewireCaptureScreen) > 0):
+        return [LEVEL_WARNING, "PipeWire capture on X11",
+                """While technically possible, most Desktop Environments do not implement the PipeWire capture portals on X11. Your captures will therefore likely not work, i.e. you will be unable to pick a window or display, or the selected source will stay empty.<br><br>
```
```suggestion
                """Most Desktop Environments do not implement the PipeWire capture portals on X11. This can result in being unable to pick a window or display, or the selected source will stay empty.<br><br>
```

mihawk90 replied:
Done.

Fenrirthviti commented on checks/linux.py:
```diff
@@ -0,0 +1,172 @@
+        return [LEVEL_WARNING, "PipeWire capture on X11",
+                """While technically possible, most Desktop Environments do not implement the PipeWire capture portals on X11. Your captures will therefore likely not work, i.e. you will be unable to pick a window or display, or the selected source will stay empty.<br><br>
+                We generally recommend using \"Window Capture (Xcomposite)\" on X11, as \"Display Capture (XSHM)\" can introduce bottlenecks depending on your setup.<br><br>
+                From a technical standpoint a Desktop Environment should not advertise the availability of these captures if they are not implemented. This behavior can therefore be considered a bug in your Desktop Environment."""]
```
This additional line doesn't feel relevant to troubleshooting, and is a personal opinion at best which we should be avoiding.

mihawk90 replied:
Yeah you're probably right, removed.

Fenrirthviti commented on checks/linux.py:
```diff
@@ -0,0 +1,172 @@
+        return [LEVEL_INFO, "Virtual Camera not available",
+                """Using the Virtual Camera requires the <code>v4l2loopback</code> kernel module to be installed.<br>
+                If required, please refer to our <a href="https://github.com/obsproject/obs-studio/wiki/install-instructions#prerequisites-for-all-versions">Install Instructions</a> on how to install this on your distribution.<br>
+                If the module was not already loaded, OBS will normally ask you for permission to load it when required. This requires a working <code>polkit</code> setup.<br>
+                You can also load the module manually using <code>modprobe v4l2loopback exclusive_caps=1 card_label='OBS Virtual Camera'</code>."""]
```
This feels a bit long-winded, and if this is a common issue, we should be updating the virtual camera troubleshooting page to include steps to resolve common Linux vcam issues and just linking that instead. Publicly indexable pages are far, far better than log analyzer messages.

mihawk90 replied:
@tiberium-v would probably know better then me, but from what I can remember it happens primarily with Arch users because they never had reason to set up `polkit`. I think most OOTB-distros come with it preinstalled these days because some system settings panels require it (e.g. setting an SDDM theme in KDE's system settings, the various partition managers, etc.).

Not sure where a good place for it would be. Maybe we could also put it in the [Prerequisites section](https://github.com/obsproject/obs-studio/wiki/install-instructions#prerequisites-for-all-versions)? Since that's where users are pointed to anyway.

Fenrirthviti replied:
We have a page for vcam stuff already, here: https://obsproject.com/kb/virtual-camera-troubleshooting so it should go there. I understand it's important to have the installation pre-requisite in the install instructions, but specific troubleshooting steps should live in the troubleshooting guide.

Feel free to just ping me in the #documentation channel on Discord with what should be added there, and I can take care of it ASAP.

mihawk90 replied:
Removed and [added to install instructions](https://github.com/obsproject/obs-studio/wiki/Install-Instructions/_compare/70e4df611cd4a982a21e9884522d62f3072a948c...6dd86287ec3440d8f064ad669ad39cbbc6912a95).

tiberium-v replied:
As a note here, it was/is Arch and also an old Debian version. In essential, it can happen on every distro, when the user switched from a traditional DE to a WM and forget to load polkit by them self.

Fenrirthviti reviewed (changes requested):
Added some verbiage changes and my opinion on the vcam stuff.

RytoEX reviewed (approved):
Going to call this good enough and scream test it.
